File: python/tn/handlers/fs_drop.py
"""``fs.drop`` handler — write admin-log snapshots to a watched directory.

Per the 2026-04-24 admin log architecture plan §5.2: useful for offline
or air-gap scenarios. The handler subscribes to admin events. On each
emit (after filtering) it composes ``tn.export(out_path, kind=
"admin_log_snapshot")`` and the resulting ``.tnpkg`` is dropped into a
local outbox directory. A peer's ``FsScanHandler`` (or any other
process) picks the file up and absorbs it.

Config shape (tn.yaml)::

    - kind: fs.drop
      out_dir: ./outbox
      on: ["tn.recipient.added", "tn.recipient.revoked"]
      scope: admin
      trigger: on_emit                    # default: on_emit
      filename_template: "snapshot_{ceremony_id}_{date}_{head_row_hash:short}.tnpkg"

Defaults:
* ``out_dir`` -> ``<yaml_dir>/.tn/outbox``
* ``trigger`` -> ``on_emit`` (FS dump on every accepted event)
* ``filename_template`` -> ``"snapshot_{ceremony_id}_{date}_{head_row_hash:short}.tnpkg"``

The ``on:`` knob is a per-event-type allowlist (an inclusive filter on
top of the ``filter:`` block). When set, only listed event types
trigger a drop. When unset, every admin event (``tn.*``) triggers.

Idempotency: the handler tracks the last-shipped ``head_row_hash`` in
memory and skips writing a duplicate file when nothing has advanced.
"""

from __future__ import annotations

import logging
import threading
from collections.abc import Callable
from datetime import datetime, timezone
from pathlib import Path
from typing import Any

from .base import TNHandler

_log = logging.getLogger("tn.handlers.fs_drop")

DEFAULT_FILENAME_TEMPLATE = "snapshot_{ceremony_id}_{date}_{head_row_hash:short}.tnpkg"


class FsDropHandler(TNHandler):
    """Drop ``.tnpkg`` admin snapshots into a local watched directory.

    Constructor parameters
    ----------------------

    name
        Handler name.
    out_dir
        Destination directory. Created on demand.
    on
        Optional list of event types to drop on (allowlist). When None,
        every admin event triggers (subject to the ``filter`` block).
    scope
        ``scope`` field passed to ``tn.export``.
    trigger
        ``"on_emit"`` (default) — drop on each accepted emit.
        ``"on_schedule"`` reserved; not implemented yet (would mirror
        ``vault.push`` scheduler).
    filename_template
        Python ``str.format`` template. Available placeholders:

        * ``{ceremony_id}``: the producer ceremony id
        * ``{date}``: ISO timestamp ``YYYYMMDDTHHMMSSZ``
        * ``{head_row_hash}``: full row_hash (or empty string)
        * ``{head_row_hash:short}``: first 12 chars after the
          ``sha256:`` prefix (or first 12 of the raw hash). Falls
          back to "noop" when the snapshot has no head.
        * ``{from_did}``: the writer's DID

    cfg_provider
        Test seam for the active LoadedConfig.
    filter_spec
        Standard handler filter dict.
    """

    def __init__(
        self,
        name: str,
        *,
        out_dir: Path,
        on: list[str] | None = None,
        scope: str = "admin",
        trigger: str = "on_emit",
        filename_template: str = DEFAULT_FILENAME_TEMPLATE,
        cfg_provider: Callable[[], Any] | None = None,
        filter_spec: dict[str, Any] | None = None,
    ) -> None:
        super().__init__(name, filter_spec)
        if trigger != "on_emit":
            # Reserved value — keep the API stable for the future.
            raise ValueError(
                f"fs.drop: trigger={trigger!r} not supported yet; only 'on_emit' is implemented."
            )
        self._out_dir = Path(out_dir)
        self._on_types: set[str] | None = set(on) if on else None
        self._scope = scope
        self._trigger = trigger
        self._filename_template = filename_template
        self._cfg_provider = cfg_provider or _default_cfg_provider

        self._last_shipped_head: str | None = None
        self._lock = threading.Lock()

    # ------------------------------------------------------------------
    # TNHandler contract
    # ------------------------------------------------------------------

    def accepts(self, envelope: dict[str, Any]) -> bool:
        if not super().accepts(envelope):
            return False
        et = envelope.get("event_type")
        if not isinstance(et, str) or not et.startswith("tn."):
            return False
        if self._on_types is not None and et not in self._on_types:
            return False
        return True

    def emit(self, envelope: dict[str, Any], raw_line: bytes) -> None:
        try:
            self._drop_snapshot(envelope)
        except Exception:  # noqa: BLE001 — preserve broad swallow; see body of handler
            _log.exception("[%s] fs.drop emit failed", self.name)

    # ------------------------------------------------------------------
    # Snapshot writer
    # ------------------------------------------------------------------

    def _drop_snapshot(self, envelope: dict[str, Any]) -> Path | None:
        cfg = self._cfg_provider()
        if cfg is None:
            _log.debug("[%s] fs.drop: no active cfg, skipping", self.name)
            return None

        from ..export import export
        from ..tnpkg import _read_manifest

        self._out_dir.mkdir(parents=True, exist_ok=True)

        # Use a temp filename first; rename only after we know head_row_hash.
        tmp_name = f"snapshot_inflight_{datetime.now(timezone.utc).strftime('%Y%m%dT%H%M%S%f')}.tnpkg"
        tmp_path = self._out_dir / tmp_name
        try:
            export(tmp_path, kind="admin_log_snapshot", cfg=cfg, scope=self._scope)
        except Exception:
            _log.exception("[%s] fs.drop: export failed", self.name)
            raise

        try:
            manifest, _body = _read_manifest(tmp_path)
        except Exception:
            _log.exception("[%s] fs.drop: failed to read back manifest", self.name)
            tmp_path.unlink(missing_ok=True)
            raise

        head = manifest.head_row_hash
        with self._lock:
            if head and head == self._last_shipped_head:
                tmp_path.unlink(missing_ok=True)
                _log.debug(
                    "[%s] fs.drop: head %s unchanged; skip", self.name, head
                )
                return None

        final_name = self._format_filename(manifest)
        final_path = self._out_dir / final_name
        if final_path.exists():
            # If the templated filename collides (e.g. two events share a
            # head_row_hash), append a microsecond suffix to keep it
            # unique. Idempotent dedupe lives in absorb on the receiver.
            ts = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%S%f")
            final_path = final_path.with_name(
                f"{final_path.stem}__{ts}{final_path.suffix}"
            )
        tmp_path.rename(final_path)

        with self._lock:
            self._last_shipped_head = head

        _log.info(
            "[%s] fs.drop: wrote snapshot %s (head=%s, trigger=%s)",
            self.name,
            final_path.name,
            head,
            envelope.get("event_type"),
        )
        return final_path

    def _format_filename(self, manifest: Any) -> str:
        head = manifest.head_row_hash or ""
        head_short = _short_hash(head)
        ts = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
        # Manual placeholder substitution because Python's str.format does
        # not support the custom ``:short`` modifier idiom we exposed in
        # the config. Keep it explicit; smaller surface than a custom
        # Formatter subclass.
        out = self._filename_template
        replacements = {
            "{ceremony_id}": manifest.ceremony_id,
            "{date}": ts,
            "{head_row_hash}": head,
            "{head_row_hash:short}": head_short,
            "{from_did}": manifest.from_did,
        }
        for needle, value in replacements.items():
            out = out.replace(needle, str(value))
        return _sanitize_filename(out)


def _short_hash(rh: str) -> str:
    if not rh:
        return "noop"
    if rh.startswith("sha256:"):
        return rh[len("sha256:") : len("sha256:") + 12]
    return rh[:12]


_DISALLOWED = '<>:"/\\|?*'


def _sanitize_filename(name: str) -> str:
    """Strip path separators and other Windows-illegal characters from
    a filename. Defensive — production uses templates that already avoid
    these, but tests sometimes feed colon-bearing DIDs through.
    """
    out = name
    for ch in _DISALLOWED:
        out = out.replace(ch, "_")
    return out


def _default_cfg_provider() -> Any:
    try:
        from .. import current_config

        return current_config()
    except RuntimeError:
        return None


__all__ = ["DEFAULT_FILENAME_TEMPLATE", "FsDropHandler"]

File: ts-sdk/src/handlers/fs_drop.ts
// `fs.drop` handler — write `.tnpkg` admin snapshots to a watched directory.
//
// Mirrors `python/tn/handlers/fs_drop.py` byte-for-byte where it matters:
// the same default filename template, the same idempotency guard
// (skip-when-head-unchanged), and the same on-emit semantics.
//
// Design notes:
// * `_buildSnapshot` is a callback rather than a hard reference to
//   `TNClient` so this module can sit underneath the TNClient/NodeRuntime
//   layer without a circular import. The host (TNClient or a test)
//   provides a builder that calls `client.export({kind:"admin_log_snapshot",...})`
//   and returns the resulting bytes plus the parsed manifest.
// * Filesystem writes are sync — these handlers don't try to overlap I/O
//   with each other; the runtime calls `emit` from a single thread.

import { existsSync, mkdirSync, readFileSync, renameSync, unlinkSync, writeFileSync } from "node:fs";
import { join } from "node:path";

import { BaseTNHandler, type FilterSpec } from "./base.js";
import { type Manifest } from "../core/tnpkg.js";
import { readTnpkg } from "../tnpkg_io.js";

/** Default filename template — must match `python/tn/handlers/fs_drop.py`. */
export const DEFAULT_FS_DROP_FILENAME_TEMPLATE =
  "snapshot_{ceremony_id}_{date}_{head_row_hash:short}.tnpkg";

/** Builder shape — the host wires this to `client.export(...)`. */
export interface SnapshotBuilder {
  /**
   * Build an admin-log snapshot at `outPath`, returning the manifest
   * bytes (file is written as a side effect) and the parsed manifest.
   */
  buildSnapshot(outPath: string, scope: string): { bytes: Uint8Array; manifest: Manifest };
}

export interface FsDropHandlerOptions {
  /** Destination directory. Created on demand. */
  outDir: string;
  /** Snapshot builder (typically a TNClient adapter). */
  builder: SnapshotBuilder;
  /** Allowlist of event types to drop on; null/undefined = every `tn.*` event. */
  on?: readonly string[];
  /** Snapshot scope — passed to export. Default `"admin"`. */
  scope?: string;
  /** Filename template; see `DEFAULT_FS_DROP_FILENAME_TEMPLATE`. */
  filenameTemplate?: string;
  filter?: FilterSpec;
}

/** Drop `.tnpkg` admin snapshots into a local watched directory. */
export class FsDropHandler extends BaseTNHandler {
  private readonly outDir: string;
  private readonly onTypes: ReadonlySet<string> | null;
  private readonly scope: string;
  private readonly filenameTemplate: string;
  private readonly builder: SnapshotBuilder;
  private lastShippedHead: string | null = null;

  constructor(name: string, opts: FsDropHandlerOptions) {
    super(name, opts.filter);
    this.outDir = opts.outDir;
    this.builder = opts.builder;
    this.onTypes = opts.on ? new Set(opts.on) : null;
    this.scope = opts.scope ?? "admin";
    this.filenameTemplate = opts.filenameTemplate ?? DEFAULT_FS_DROP_FILENAME_TEMPLATE;
  }

  override accepts(envelope: Record<string, unknown>): boolean {
    if (!super.accepts(envelope)) return false;
    const et = String(envelope["event_type"] ?? "");
    if (!et.startsWith("tn.")) return false;
    if (this.onTypes !== null && !this.onTypes.has(et)) return false;
    return true;
  }

  emit(_envelope: Record<string, unknown>, _rawLine: string): void {
    try {
      this.dropSnapshot();
    } catch (e) {
      console.warn(`[${this.name}] fs.drop emit failed:`, e);
    }
  }

  /** Public test seam — synchronously build + write one snapshot. */
  dropSnapshot(): string | null {
    if (!existsSync(this.outDir)) mkdirSync(this.outDir, { recursive: true });
    const tmpStamp = nowStampMicro();
    const tmpPath = join(this.outDir, `snapshot_inflight_${tmpStamp}.tnpkg`);

    let manifest: Manifest;
    try {
      const built = this.builder.buildSnapshot(tmpPath, this.scope);
      manifest = built.manifest;
    } catch (e) {
      try {
        unlinkSync(tmpPath);
      } catch {
        // best-effort
      }
      throw e;
    }

    const head = manifest.headRowHash ?? null;
    if (head !== null && this.lastShippedHead === head) {
      try {
        unlinkSync(tmpPath);
      } catch {
        // best-effort
      }
      return null;
    }

    const finalName = formatFilename(
      this.filenameTemplate,
      manifest.ceremonyId,
      head,
      manifest.fromDid,
    );
    let finalPath = join(this.outDir, finalName);
    if (existsSync(finalPath)) {
      const suffix = nowStampMicro();
      const dot = finalName.lastIndexOf(".");
      const stem = dot >= 0 ? finalName.slice(0, dot) : finalName;
      const ext = dot >= 0 ? finalName.slice(dot) : "";
      finalPath = join(this.outDir, `${stem}__${suffix}${ext}`);
    }
    renameSync(tmpPath, finalPath);
    this.lastShippedHead = head;
    return finalPath;
  }
}

/** Helper used by tests/hosts to build a snapshot via TNClient. */
export function makePackageSnapshotBuilder(client: {
  export: (opts: { kind: string; scope?: string }, outPath: string) => string;
}): SnapshotBuilder {
  return {
    buildSnapshot(outPath: string, scope: string) {
      client.export({ kind: "admin_log_snapshot", scope }, outPath);
      const bytes = readFileSync(outPath);
      const { manifest } = readTnpkg(bytes);
      return { bytes, manifest };
    },
  };
}

const DISALLOWED = /[<>:"/\\|?*]/g;

function sanitizeFilename(name: string): string {
  return name.replace(DISALLOWED, "_");
}

function shortHash(rh: string | null): string {
  if (rh === null || rh === "") return "noop";
  const stripped = rh.startsWith("sha256:") ? rh.slice("sha256:".length) : rh;
  return stripped.slice(0, 12);
}

function nowStampMicro(): string {
  // Python uses %Y%m%dT%H%M%S%f (microseconds). JS Date only has ms; pad
  // with three zeros for parity in shape, not in precision.
  const d = new Date();
  const pad = (n: number, w: number) => String(n).padStart(w, "0");
  return (
    `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1, 2)}${pad(d.getUTCDate(), 2)}` +
    `T${pad(d.getUTCHours(), 2)}${pad(d.getUTCMinutes(), 2)}${pad(d.getUTCSeconds(), 2)}` +
    `${pad(d.getUTCMilliseconds(), 3)}000`
  );
}

function nowStampSeconds(): string {
  const d = new Date();
  const pad = (n: number, w: number) => String(n).padStart(w, "0");
  return (
    `${d.getUTCFullYear()}${pad(d.getUTCMonth() + 1, 2)}${pad(d.getUTCDate(), 2)}` +
    `T${pad(d.getUTCHours(), 2)}${pad(d.getUTCMinutes(), 2)}${pad(d.getUTCSeconds(), 2)}Z`
  );
}

/** Substitute placeholders in a fs.drop filename template. */
export function formatFilename(
  template: string,
  ceremonyId: string,
  headRowHash: string | null,
  fromDid: string,
): string {
  const head = headRowHash ?? "";
  const headShort = shortHash(headRowHash);
  const date = nowStampSeconds();
  // Order matters: replace the longer placeholder first so it doesn't get
  // partially eaten by the shorter one.
  let out = template;
  out = out.replace("{head_row_hash:short}", headShort);
  out = out.replace("{head_row_hash}", head);
  out = out.replace("{ceremony_id}", ceremonyId);
  out = out.replace("{date}", date);
  out = out.replace("{from_did}", fromDid);
  return sanitizeFilename(out);
}

/** Used by `loadHandlers` to instantiate from yaml. */
export interface FsDropSpec {
  kind: "fs.drop";
  name?: string;
  out_dir?: string;
  on?: readonly string[];
  scope?: string;
  trigger?: string;
  filename_template?: string;
  filter?: FilterSpec;
}

void writeFileSync;
